requires.py

```diff
@@ -45,13 +61,28 @@ def services(self):
             })
             for unit in relation.joined_units:
                 data = unit.received_raw
-                private_address = data['private-address']
-                host = data['hostname'] or private_address
-                port = data['port']
-                if host and port:
-                    service['hosts'].append({
-                        'hostname': host,
-                        'private-address': private_address,
-                        'port': port,
-                    })
-        return [s for s in services.values() if s['hosts']]
+                host = build_service_host(data)
+                if host:
+                    # remove duplicates
+                    key = '{}.{}.{}'.format(host['hostname'],
+                                            host['private-address'],
+                                            host['port'])
+                    if key not in dup_list:
+                        dup_list[key] = 1
+                        service['hosts'].append(host)
```

Comment:
Python [sets](https://docs.python.org/3.7/library/stdtypes.html#set-types-set-frozenset) could be a cleaner way to handle deduplication here. For example:

```
>>> hosts = set()  # set of tuples (hostname, private_address, port)
>>> hosts.add(('111.111.111.111', '111.111.111.111', 111))
>>> hosts.add(('222.222.222.222', '222.222.222.222', 222))
>>> hosts.add(('111.111.111.111', '111.111.111.111', 111))  # already in set, no-op
>>> for hostname, private_address, port in hosts:
...     print(hostname, private_address, port)
... 
111.111.111.111 111.111.111.111 111
222.222.222.222 222.222.222.222 222
```